### ISSUE TITLE
MPIX_Query_cuda_support triggers assertion error

### ISSUE THREAD

**Issue by simonbyrne** [pmodels/mpich#5716]
Calling `MPIX_Query_cuda_support()`  on a Mac triggers an assertion error:
```
Assertion failed: (mpi_errno), function MPIX_Query_cuda_support, file src/mpi/misc/gpu.c, line 90.
```
Seen in both the homebrew version, and the Julia BinaryBuilder-built one

**Comment by hzhou** [comment]
Did you call `MPIX_Query_cuda_support()` after `MPI_Init`?

**Comment by simonbyrne** [comment]
Apologies, I was incorrect: the homebrew one is fine, it only affects the Julia BinaryBuilder one. I have opened an issue there (https://github.com/JuliaPackaging/Yggdrasil/issues/4039), but if you have any suggestions for what to look for, the build script is here:
https://github.com/JuliaPackaging/Yggdrasil/blob/a6ad87971cc8bc11de301cc3e53cb8bfb4d5bcfb/M/MPICH/build_tarballs.jl

**Comment by simonbyrne** [comment]
Actually, I think the issue is that it is built with debug enabled, which enables assertions: there is an assertion on the error code returned by `PMPIX_GPU_query_support`:
https://github.com/pmodels/mpich/blob/4e5ce994675370015145f50f4eff5efc3a2e5aa8/src/binding/c/misc_api.txt#L173-L174
i.e. it is asserting a non-zero return value, but `MPI_SUCCESS == 0`. Should this be changed to
```C
assert(mpi_errno == MPI_SUCCESS);
```
?

**Comment by hzhou** [comment]
> Actually, I think the issue is that it is built with debug enabled, which enables assertions: there is an assertion on the error code returned by `PMPIX_GPU_query_support`:
> 
> https://github.com/pmodels/mpich/blob/4e5ce994675370015145f50f4eff5efc3a2e5aa8/src/binding/c/misc_api.txt#L173-L174
> 
> i.e. it is asserting a non-zero return value, but `MPI_SUCCESS == 0`. Should this be changed to
> 
> ```c
> assert(mpi_errno == MPI_SUCCESS);
> ```
> 
> ?

You are absolutely correct!

**Comment by giordano** [comment]
Am I correct to say this isn't fixed in the just released v3.4.3?  In the [`v3.4.3`](https://github.com/pmodels/mpich/tree/v3.4.3) tag I still see
https://github.com/pmodels/mpich/blob/7aeecca034b145d82a35f9f3cad20534e792ac6c/src/mpi/misc/gpu.c#L90

**Comment by raffenet** [comment]
> Am I correct to say this isn't fixed in the just released v3.4.3? In the [`v3.4.3`](https://github.com/pmodels/mpich/tree/v3.4.3) tag I still see
> 
> https://github.com/pmodels/mpich/blob/7aeecca034b145d82a35f9f3cad20534e792ac6c/src/mpi/misc/gpu.c#L90

Correct, it is not fixed in 3.4.3. The fix will be part of a soon to be released 4.0rc1 (and new releases). Apologies for the delay.

**Comment by ax3l** [comment]
Ouch, Perlmutter (NERSC) still uses this MPICH version and I am hitting it now with 3.4.3 😅  (inside a container that is intentionally GPU-unaware... don't ask why, I am two layers down in work-arounds.)
